app/javascript/components/student/editor/TestRunSummary.tsx

```diff
@@ -57,40 +49,51 @@ export function TestRunSummary({
   submission: Submission
   timeout: number
 }) {
-  const RESOLVED_TEST_STATUSES = [
-    TestRunStatus.PASS,
-    TestRunStatus.FAIL,
-    TestRunStatus.ERROR,
-    TestRunStatus.OPS_ERROR,
-  ]
-  const [testRun, dispatch] = useReducer(reducer, {
+  const [testRun, setTestRun] = useState<TestRun>({
     submissionUuid: submission.uuid,
     status: submission.testsStatus,
     message: '',
     tests: [],
   })
-  const haveTestsResolved = RESOLVED_TEST_STATUSES.includes(testRun.status)
   const timer = useRef<number>()
+  const channel = useRef<TestRunChannel>()
 
   useEffect(() => {
-    const channel = new TestRunChannel(submission, (testRun: TestRun) => {
-      dispatch({ type: 'testRun.received', payload: testRun })
+    channel.current = new TestRunChannel(submission, (testRun: TestRun) => {
+      setTestRun(testRun)
     })
 
     return () => {
-      channel.disconnect()
+      if (channel.current) {
+        channel.current.disconnect()
+      }
     }
   }, [submission.uuid])
 
   useEffect(() => {
-    if (haveTestsResolved) {
-      return
-    }
+    switch (testRun.status) {
+      case TestRunStatus.QUEUED:
+        timer.current = window.setTimeout(
+          () => setTestRun({ ...testRun, status: TestRunStatus.TIMEOUT }),
+          timeout
+        )
 
-    timer.current = window.setTimeout(
-      () => dispatch({ type: 'testRun.timeout' }),
-      timeout
-    )
+        break
+      case TestRunStatus.CANCELLING:
+        fetchJSON(submission.links.cancel, {
+          method: 'POST',
+        }).then(() => {
+          setTestRun({ ...testRun, status: TestRunStatus.CANCELLED })
+        })
+
+        break
+      case TestRunStatus.CANCELLED:
+        if (channel.current) {
+          channel.current.disconnect()
+        }
+
+        break
```

Comment:
Just a small suggestion: it might help readability to introduce some helper functions here. That way you could do something like:

```
switch (testRun.status) {
  case TestRunStatus.QUEUED:
    handleQueued()
    break
  case TestRunStatus.CANCELLING:
    handleCancelling()
    break
  case TestRunStatus.CANCELLED:
    handleCancelled()
    break
}
```

But feel free to ignore this if you think this does not help.

Reply:
I like this suggestion a lot.